python/ray/tune/execution/tune_controller.py

```diff
@@ -0,0 +1,1196 @@
+    def _maybe_add_actors(self):
+        """Add actors for pending and paused trials.
+
+        For actors that have not been staged, yet, we request an actor.
+
+        For actors that have been staged, already, we try to reuse a cached actor.
+
+        First, we handle the trial that the scheduler chooses to run.
+
+        Then, we handle all trials that are pending or paused.
+
+        Lastly, we see if we have cached actors that we can assign to a pending or
+        paused trial. This can be the case when a trial has not been staged, yet,
+        for instance because the number of staging trials is too large.
+        """
+
+        ###
+        # 1: Start trial that the scheduler wants to run
+        with warn_if_slow("choose_trial_to_run"):
+            trial_to_run = self._scheduler_alg.choose_trial_to_run(self._wrapped())
+
+        if trial_to_run:
+            logger.debug(f"Chose trial to run from scheduler: {trial_to_run}")
+            if (
+                trial_to_run not in self._staged_trials
+                and trial_to_run not in self._trial_to_actor
+            ):
+                logger.debug(f"Staging trial to run: {trial_to_run}")
+                self._staged_trials.add(trial_to_run)
+                self._actor_cache.increase_max(trial_to_run.placement_group_factory)
+                # schedule_trial_actor also potentially uses cached actors
+                self._schedule_trial_actor(trial_to_run)
+            else:
+                # Otherwise, only try to use the cached actor
+                logger.debug(f"Trying to re-use actor for trial to run: {trial_to_run}")
+                self._maybe_reuse_cached_actor(trial_to_run)
+
+        ###
+        # 2: Start trials that are PENDING or PAUSED
+        def _maybe_add_actors(candidates: List[Trial]):
+            new_candidates = []
+
+            while candidates:
+                if len(self._staged_trials) >= self._max_pending_trials:
+                    break
+
+                trial = candidates.pop(0)
+
+                # If the trial is part of the list, but not of the set,
+                # we just ignore it. Removing it from the list on status
+                # change is too expensive.
+                if trial not in (self._pending_trials | self._paused_trials):
+                    continue
+
+                if trial in self._trial_to_actor:
+                    new_candidates.append(trial)
+                    continue
+
+                if trial in self._staged_trials:
+                    self._maybe_reuse_cached_actor(trial)
+                    continue
+
+                logger.debug(f"Scheduling actor for enqueued trial: {trial}")
+                self._staged_trials.add(trial)
+                self._actor_cache.increase_max(trial.placement_group_factory)
+                self._schedule_trial_actor(trial)
+
+            return new_candidates + candidates
+
+        self._pending_trials_list = _maybe_add_actors(self._pending_trials_list)
```

Comment:
why treat pending and paused separately?

Reply:
Actually this was a regression - paused trials should not be handled here, as they should be selected by the scheduler instead (otherwise things like synchronous PBT won't work). Removed the path for paused trials!